File: offset-manager-api/src/test/java/offsetmanager/domain/file/factory/FileKeyParserTest.java
package offsetmanager.domain.file.factory;

import offsetmanager.domain.file.FileKey;
import offsetmanager.domain.file.LocalFileKey;
import offsetmanager.domain.file.S3FileKey;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.*;

class FileKeyParserTest {

  @DisplayName("LocalFileKey should be created when file URI as local file path")
  @Test
  void localFileKeyPathTest() {
    FileKey fileKey = FileKeyParser.parse("file:///path/to/file.txt");
    assertThat(fileKey).isInstanceOf(LocalFileKey.class);
  }

  @DisplayName("S3FileKey should be created when S3 Uri provided")
  @Test
  void S3FileKeyPathTest() {
    FileKey fileKey = FileKeyParser.parse("s3://my-bucket/path/to/file.txt");
    assertThat(fileKey).isInstanceOf(S3FileKey.class);
  }

  @DisplayName("Should throw IllegalArgumentException when unsupported URI scheme provided")
  @Test
  void invalidFileKeyPathTest() {
    assertThatThrownBy(()-> FileKeyParser.parse("invalid file key string"))
      .isInstanceOf(IllegalArgumentException.class)
      .hasMessage("Unsupported file key schema: " + "invalid file key string");
  }
}

File: offset-manager-api/src/test/kotlin/offsetmanager/domain/file/factory/FileKeyParserTest.kt
package offsetmanager.domain.file.factory

import offsetmanager.domain.file.LocalFileKey
import offsetmanager.domain.file.S3FileKey
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.assertj.core.api.AssertionsForInterfaceTypes.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

internal class FileKeyParserTest {
  @DisplayName("Windows C drive style path local file key toUri creation test")
  @Test
  fun windowsLocalFIleKeyPathTest() {
    val fileKey = FileKeyParser.parse("file:///C:/path/to/file.txt")
    assertThat(fileKey).isInstanceOf(LocalFileKey::class.java)
    assertThat(fileKey.get()).isEqualTo("file:///C:/path/to/file.txt")
  }

  @DisplayName("LocalFileKey should be created when file URI as local file path")
  @Test
  fun localFileKeyPathTest() {
    val fileKey = FileKeyParser.parse("file:///path/to/file.txt")
    assertThat(fileKey).isInstanceOf(LocalFileKey::class.java)
    assertThat(fileKey.get()).isEqualTo("file:///path/to/file.txt")
  }

  @DisplayName("Should throw IllegalArgumentException when schema is missing")
  @Test
  fun schemaMissingTest() {
    // given
    assertThatThrownBy({ FileKeyParser.parse("/path/to/file.txt") })
      .isInstanceOf(IllegalArgumentException::class.java)
      .hasMessage("Unsupported file key schema: " + "/path/to/file.txt")
  }

  @DisplayName("S3FileKey should be created when S3 Uri provided")
  @Test
  fun s3FileKeyPathTest() {
    val fileKey = FileKeyParser.parse("s3://my-bucket/path/to/file.txt")
    assertThat(fileKey).isInstanceOf(S3FileKey::class.java)
  }
}
